cluster-autoscaler/proposals/provisioning-request.md

```diff
@@ -0,0 +1,289 @@
+}
+
+type PodSet struct {
+	// TemplateRef is a reference to a PodTemplate object that is representing pods 
+	// that will consume this reservation. Users need to make sure that the
+	// fields relevant to scheduler (e.g. node selector tolerations) are
+	// consistent between this template and actual pods consuming the Provisioning Request.
+	TemplateRef *metav1.ObjectReference `json:"templateRef" protobuf:"bytes,1,name=templateRef"`
+	// Count contains the number of pods that will be created with a given
+	// template.
+	// +kubebuilder:validation:Minimum=1
+	Count int32 `json:"count" protobuf:"bytes,2,name=count"`
+}
+
+// ProvisioningRequestStatus represents the status of the resource reservation.
+type ProvisioningRequestStatus struct {
+	// Conditions represent the observations of a Provisioning Request's
+	// current state. Those will contain information whether the capacity
+    // was found/created or if there were any issues. The condition types
+    // may differ between different provisioning classes.
+	//
+	// +optional
+	Conditions []metav1.Condition `json:"conditions" protobuf:"bytes,1,rep,name=conditions"`
+
+	// AdditionalStatus contains all other status values custom provisioning classes may require.
+	//
+	// +optional
+	AdditionalStatus map[string]string `json:"additionalStatus" protobuf:"bytes,2,name=additionalStatus"`
+}
+```
+
+### Provisioning Classes
+
+#### kubernetes.io/checkCapacity class
+
+The `kubernetes.io/checkCapacity` is one-off check to verify that the in the cluster
+there is enough capacity to provision given set of pods.
+
+Note: If two of such objects are created around the same time, CA will consider
+them independently and place no guards for the capacity.
+Also the capacity is not reserved in any manner so it may be scaled-down.
+
+#### kubernetes.io/atomicScaleUp class
+
+The `kubernetes.io/atomicScaleUp` aims to provision the resources required for the
+specified pods in an atomic way. The proposed logic is to:
+1. Try to provision required VMs in one loop.
+2. If it failed, remove the partially provisioned VMs and back-off.
+3. Stop the back-off after a given duration (optional), which would be passed
+   via `AdditionalParameters` field, using `ValidUntilSeconds` key and would contain string
+   denoting duration for which we should retry (measured since creation fo the CR).
+
+Note: that the VMs created in this mode are subject to the scale-down logic.
+So the duration during which users need to create the Pods is equal to the
+value of `--scale-down-unneeded-time` flag.
+
+### Adding pods that consume given ProvisioningRequest
+
+To avoid generating double scale-ups and exclude pods that are meant to consume
+given capacity CA should be able to differentiate those from all other pods.  
+To do so users need to specify the following pod annotation (it is not required
+in ProvReq’s template, though it can be specified):
+
+```yaml
+annotations:
```

Comment:
Is it possible for CAS to observe that there is other capacity that is available and managed by the the ProvisioningRequest and, therefore, not scale-up for a given set of pods that can fit inside of that requested capacity?

Reply:
This annotation is meant to avoid double counting of the needed capacity, i.e. the actual pods and the ones inferred from Provisioning Request.
Which this CA knows that for those VMs there was a dedicated capacity managed by the Provisioning Request.